content/blog/warnings.md

```diff
@@ -0,0 +1,173 @@
+```
+
+Sometimes these "helpful" warn-and-continue effects continue as the NA or zero-length vector created wrecks havoc on your code leading to "missing value where TRUE/FALSE needed" or "argument is of length zero" type errors where they finally hit an `if` statement, with the actual problem now quite distant from the solution.
+
+Aha, but you can turn warnings into errors, and then use normal error handling to investigate?
+
+```r
+options(warn = 2)
+f("two")
+#> Error in rnorm(n) : (converted from warning) NAs introduced by coercion
+```
+
+which is fine if your warning is the *first* warning to occur.  But if your code is littered with "informational warnings" then the error will be thrown before your new error is reached.
+
+By default R only shows the last 10 warnings, and only keeps hold of the last 50, so it's not guaranteed you'll be able to read the warning that really caused the problem.
+
+## The problem
+
+Base R is filled with cases where warnings are used where errors would have been more appropriate:
+
+```r
+readRDS("nosuchfile")
+#> Error in gzfile(file, "rb") : cannot open the connection
+#> In addition: Warning message:
+#> In gzfile(file, "rb") :
+#>   cannot open compressed file 'nosuchfile', probable reason 'No such file or directory'
+```
+
+Here, `gzfile` tries to open a file which doesn't exist, "throws" a warning, then just lets the program continue, passing back a connection object that cannot be read from. When something does read from it, it ends badly. However, the true error (the file does not exist) is obscured as a warning and the user is left holding a weirder one ("cannot open a connection").
+
+Once you add in the maximum number of warnings issue, things get more annoying.  Consider a completely silly example:
+
+```r
+local({
+  for (i in 1:50) {
+    Sys.timezone(FALSE)
+  }
+  readRDS("nosuchfile")
+})
+#> Error in gzfile(file, "rb") : cannot open the connection
+#> In addition: There were 50 or more warnings (use warnings() to see the first 50)
+```
+
+Similar to above, except we do not get the real error here anymore. Worse, when following the instructions and running `warnings()` our true error is not found as every warning printed is about a defunct argument to `Sys.timezone`
+
+```r
+#> 1: Sys.timezone(location = FALSE) is defunct and ignored
+#> 2: Sys.timezone(location = FALSE) is defunct and ignored
+...
+#> 50: Sys.timezone(location = FALSE) is defunct and ignored
+```
+
+This means that the catastrophic warning produced by `readRDS` is swamped by the informational one by `Sys.timezone` and the user has limited information with which to start debugging.
+
+The above example is obviously contrived but it's not that hard to accumulate 50 warnings - many packages are quite chatty with warnings as many packages use them for informational messages. Alternatively, you might have `options(warnPartialMatchDollar = TRUE)` set so that partial matching produces warnings and have used code like
+
```

Comment:
Philosophical ramblings, all in the same direction of the blog...

I've always felt (having lost some days with it the first time I came across it) that the warnPartialMatchDollar option was the "wrong kind of fix" - the speculative "auto-completing" of code seems to me a spectacularly bad idea that should just go away and be repented of, forcing authors to actually write syntactically valid code like they would in all the other languages I can think of. It's one of those things that has encouraged some very poor practice, and is difficult to rein it back in... 

...such as DBI (or is it RPostgresql?) - who still haven't merged your fix - perhaps they think "huh, it's only a warning". Hence latest version still throws warnings about partial matching, which means we have to write little wrapper functions that check/switch options(warn) on/off in some way, so that we ignore DBI's warnings, but treat others as errors.

Reply:
I think we might write another post on the dangers of dollar partial matching

Reply:
We can call it "Can R Eliminate Ambiguous Matches (or [C.R.E.A.M. - Dolla Dolla Bill Y'all](https://www.youtube.com/watch?v=PBwAxmrE194))